### ISSUE TITLE
Sagas not running

### ISSUE THREAD

**Issue by patrick-jones** [keajs/kea#32]
Given a simple kea logic definition applied to a component:

```js
const ImportQuoteLogic = kea({
 actions: () => ({
    selectFile: (fileObj: File) => ({fileObj}),
    cancelImport: () => ({}),
  }),

  reducers: ({actions}) => ({
    quoteImport: [{}, PropTypes.object, {
      [actions.selectFile]: (state, {fileObj}) => ({filename: fileObj.name}),
      [actions.cancelImport]: state => ({}),
    }],
  }),

  workers: {
    handleFile: function* (action) {
      console.log('process file');
      yield call(console.log, 'process file', action.payload);
    },
  },

  start: function* () {
    console.log('start');
    yield call(console.log, 'start');
  },

  takeEvery: ({actions, workers}) => {
    console.log('returning action to worker mapping...');
    return ({
      [actions.selectFile]: workers.handleFile,
    });
  },
});
```
The sagas don't run. `start` never prints to the console; neither does 'returning action to worker mapping'. The actions and reducers all work (as observed in the redux dev tools); just the sagas are ignored.

I have a few other kea-fied components that are behaving just fine, so it is not my overall kea setup. I feel like it must be just staring me in the face, but I've been at this for hours now and I'm getting nowhere.

**Comment by mariusandra** [comment]
Hey, can you give an example on how and from where you call the `selectFile` action?

**Comment by patrick-jones** [comment]
First of all, congratulations on the birth of your new daughter! Do yourself a favor and concentrate on her for a while.

I have finally figured out what the problem is. My component is a stateless functional component, which the docs say is supported. But sagas are started and stopped in the enhanced `componentDidMount` and `componentWillUnmount` methods that kea adds to the class prototype. But react still sees the component as a stateless functional component, and the lifecycle methods never get called. The only solution I can think of is to wrap stateless functional components with a class-based wrapper rather than replacing the lifecycle methods on the prototype.

This is a stripped down version of what I have in my project, but it still exhibits the problematic behavior. The `ImportQuote` component is rendered as a `react-router` (v4) `Route`.  Redux dev tools shows that the action fires and the state is changed, but none of the saga activity occurs. (For example, none of the console log messages are ever shown).

```js
// logic.js
import PropTypes from 'prop-types';
import {call} from 'redux-saga/effects';

import { kea } from 'kea';


const ImportQuoteLogic = kea({
  path: () => ['scenes', 'quotes', 'ImportQuote'],

  actions: () => ({
    selectFile: (fileObj: File) => ({fileObj}),
    cancelImport: () => ({}),
  }),

  reducers: ({actions}) => ({
    quoteImport: [{}, PropTypes.object, {
      [actions.selectFile]: (state, {fileObj}) => ({filename: fileObj.name}),
      [actions.cancelImport]: state => ({}),
    }],
  }),

  workers: {
    handleFile: function* (action) {
      console.log('process file', action);
      yield call(console.log, 'process file', action);
    },
  },

  start: function* () {
    console.log('start');
    yield call(console.log, 'start');
  },

  takeEvery: ({actions, workers}) => {
    console.log('returning action to worker mapping...');
    return ({
      [actions.selectFile]: workers.handleFile,
    });
  },

});

export {
  ImportQuoteLogic,
};
```

```js
// ImportQuote.js
import React from 'react';
import {ImportQuoteLogic} from './logic';

const ImportQuote = (props: Props) => {
  const {actions} = props;
  const handleChange = e => {
    const files = e.target.files;
    if (files.length > 0) {
      actions.selectFile(files[0]);
    }
  };
  return (
    <div className='c p'>
      <input
        type='file'
        onInput={handleChange}
        onChange={handleChange}
      />
    </div>
  );
};

export default ImportQuoteLogic(ImportQuote);
```

**Comment by mariusandra** [comment]
Hey, thanks for your wishes and also thanks for finding the cause of the bug! :)

I made a patch version, `0.23.1`, where I believe the bug is fixed! I made a workaround so that if you're wrapping a logic around a functional component, the sagas will not be patched to this function, but to redux's Connect component that will wrap it. I also added tests to make sure this will keep working.

Let me know if this fixes it for you! :)
